traits/trait_set_object.py

```diff
@@ -12,247 +12,425 @@
-                if validate is not None:
-                    name = self.name
-                    added = set(
-                        [validate(object, name, item) for item in added]
-                    )
-
-                set.update(self, added)
-
-                if self.name_items is not None:
-                    self._send_trait_items_event(
-                        self.name_items, TraitSetEvent(None, added)
-                    )
-        except TraitError as excp:
-            excp.set_prefix("Each element of the")
-            raise excp
+        Notifiers are transient and should not be serialized.
+        """
+        result = self.__dict__.copy()
+        # notifiers are transient and should not be serialized
+        result.pop("notifiers", None)
+        return result
 
-    def intersection_update(self, value):
-        removed = self.difference(value)
-        if len(removed) > 0:
-            set.difference_update(self, removed)
+    def __setstate__(self, state):
+        """ Restore the state of the object after serialization.
+
+        Notifiers are transient and are restored to the empty list.
+        """
+        state['notifiers'] = []
+        self.__dict__.update(state)
+
+    def add(self, value):
+        """ Add an element to a set.
+
+        This has no effect if the element is already present.
+
+        Parameters
+        ----------
+        value : any
+            The value to add to the set.
+
+        Notes
+        -----
+        Parameters in the notification:
+            removed : set
+                An empty set.
+            added : set
+                A set containing the added value.
 
-            if self.name_items is not None:
-                self._send_trait_items_event(
-                    self.name_items, TraitSetEvent(removed)
-                )
+        """
+        validated_values = self.validate(set([value]))
+        if len(validated_values) > 1:
+            raise ValueError("Validator returned {} values "
+                             "where 1 value is "
+                             "expected".format(len(validated_values)))
+
+        value = next(iter(validated_values))
+        if value not in self:
+            super().add(value)
+            self.notify(set(), validated_values)
 
-    def difference_update(self, value):
+    def remove(self, value):
+        """ Remove an element from a set; it must be a member.
+
+        If the element is not a member, raise a KeyError.
+
+        Parameters
+        ----------
+        value : any
+            An element in the set
+
+        Raises
+        ------
+        KeyError
+            If the value is not found in the set.
+
+        Notes
+        -----
+        Parameters in the notification:
+            removed : set
+                A set containing the removed value.
+            added : set
+                An empty set.
+
+        """
+        super().remove(value)
```

Comment:
There's a tricky problem to solve here. Ideally we should report the element that was actually removed from the set, rather than the argument to the `remove` call. E.g., in the snippet below, I'd really want to see `{2}` rather than `{2.0}` for the removed items.
```
Python 3.8.2 (default, Feb 27 2020, 19:56:42) 
[Clang 10.0.1 (clang-1001.0.46.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from traits.api import *
>>> class A(HasTraits):
...     foo = Set(Int)
...     def _foo_items_changed(self, event):
...         print(f"Added: {event.added}; Removed: {event.removed}")
... 
>>> a = A(foo={1, 2, 3})
>>> a.foo.remove(2.0)
Added: set(); Removed: {2.0}   # Expect {2} rather than {2.0}
```

But I don't know of any way to do this that's clean, reliable and efficient. (The core Python philosophy here is that if the two elements are equal, they should be interchangeable, so you shouldn't care.)

Reply:
See the thread starting at https://mail.python.org/pipermail/python-ideas/2013-September/023143.html for some related discussion.

Reply:
I can only think of an inefficient way:
```
>>> s = {1,2,3}
>>> removed = s.difference(s.difference({2.0}))
>>> removed
{2}
```

Reply:
Yes, if we're okay with inefficient then we can use something obvious like `{x for x in my_set if x == 2.0}` to pick out the relevant element. There's an evil trick efficient way that involves wrapping the target element (in this case `2.0`) in a custom object and then observing the calls to `__eq__` made by the Python set machinery on that custom object, but that fails the "clean" test.

This is probably a case where practicality beats purity, and we should just document the deficiency in the docstring.

Reply:
Here's some code for the non-clean efficient solution:
```python
class _EqObserver:
    def __init__(self, value):
        self.value = value
        self.matches = []

    def __eq__(self, other):
        equal = (self.value == other)
        if equal:
            self.matches.append(other)
        return equal

    def __hash__(self):
        return hash(self.value)


def set_get(s, obj):
    """
    Look "obj" up in set s.

    Return the actual element of s matching obj if it exists.
    Raise ValueError otherwise.

    Examples
    --------
    >>> set_get({1, 2, 3}, 2.0)
    2
    >>> set_get({1, 2, 3}, 4.0)
    Traceback (most recent call last):
        ...
    ValueError: object 4.0 not found in set

    """
    observer = _EqObserver(obj)
    if observer in s:
        return observer.matches[0]
    else:
        raise ValueError(f"object {obj} not found in set")
```

Reply:
Decided to not do this at the moment in an offline discussion.